File: C++/398.random-pick-index.cpp
/*
 * @lc app=leetcode id=398 lang=cpp
 *
 * [398] Random Pick Index
 *
 * https://leetcode.com/problems/random-pick-index/description/
 *
 * algorithms
 * Medium (53.15%)
 * Likes:    368
 * Dislikes: 593
 * Total Accepted:    73.5K
 * Total Submissions: 137.5K
 * Testcase Example:  '["Solution","pick"]\n[[[1,2,3,3,3]],[3]]'
 *
 * Given an array of integers with possible duplicates, randomly output the
 * index of a given target number. You can assume that the given target number
 * must exist in the array.
 * 
 * Note:
 * The array size can be very large. Solution that uses too much extra space
 * will not pass the judge.
 * 
 * Example:
 * 
 * 
 * int[] nums = new int[] {1,2,3,3,3};
 * Solution solution = new Solution(nums);
 * 
 * // pick(3) should return either index 2, 3, or 4 randomly. Each index should
 * have equal probability of returning.
 * solution.pick(3);
 * 
 * // pick(1) should return 0. Since in the array only nums[0] is equal to 1.
 * solution.pick(1);
 * 
 * 
 */
#include<vector>
using namespace std;

// @lc code=start
class Solution {
private:
    vector<int> nums;
public:
    Solution(vector<int>& nums) : nums(nums) {}
    
    int pick(int target) {
        int cnt = 0, ans = -1;
        
        for (int i=0; i<nums.size(); ++i) {
            if (nums[i] != target) continue;
            
            ++cnt;
            if (rand() % cnt == 0) ans = i;
        }
        return ans;
    }
};


/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * int param_1 = obj->pick(target);
 */
// @lc code=end

File: Python/398.random-pick-index.py
#
# @lc app=leetcode id=398 lang=python3
#
# [398] Random Pick Index
#
# https://leetcode.com/problems/random-pick-index/description/
#
# algorithms
# Medium (63.70%)
# Likes:    1291
# Dislikes: 1287
# Total Accepted:    235.9K
# Total Submissions: 370.3K
# Testcase Example:  '["Solution","pick","pick","pick"]\n[[[1,2,3,3,3]],[3],[1],[3]]'
#
# Given an integer array nums with possible duplicates, randomly output the
# index of a given target number. You can assume that the given target number
# must exist in the array.
# 
# Implement the Solution class:
# 
# 
# Solution(int[] nums) Initializes the object with the array nums.
# int pick(int target) Picks a random index i from nums where nums[i] ==
# target. If there are multiple valid i's, then each index should have an equal
# probability of returning.
# 
# 
# 
# Example 1:
# 
# 
# Input
# ["Solution", "pick", "pick", "pick"]
# [[[1, 2, 3, 3, 3]], [3], [1], [3]]
# Output
# [null, 4, 0, 2]
# 
# Explanation
# Solution solution = new Solution([1, 2, 3, 3, 3]);
# solution.pick(3); // It should return either index 2, 3, or 4 randomly. Each
# index should have equal probability of returning.
# solution.pick(1); // It should return 0. Since in the array only nums[0] is
# equal to 1.
# solution.pick(3); // It should return either index 2, 3, or 4 randomly. Each
# index should have equal probability of returning.
# 
# 
# 
# Constraints:
# 
# 
# 1 <= nums.length <= 2 * 10^4
# -2^31 <= nums[i] <= 2^31 - 1
# target is an integer from nums.
# At most 10^4 calls will be made to pick.
# 
# 
#

# @lc code=start
class Solution:

    def __init__(self, nums: List[int]):
        self.nums = nums
        self.ones = [i for i, num in enumerate(nums) if num == 1]

    def pick(self, target: int) -> int:
        if target == 1:
            return self.ones[random.randint(0, len(self.ones) - 1)]
        cnt = 0
        ans = -1
        for i, num in enumerate(self.nums):
            if num == target:
                cnt += 1
                if random.randint(1, cnt) == 1:
                    ans = i
        return ans



# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.pick(target)
# @lc code=end

